blog/content/post/2020-12-08-IOMMU.md

```diff
@@ -0,0 +1,162 @@
+that connects a Direct Memory Access (DMA) capable expansion bus to the main memory.
+It extends the system architecture by adding support for the virtualization of
+memory addresses used by peripheral devices. Additionally, it provides memory
+isolation and protection by enabling system software to control which areas of
+physical memory an I/O device may access. It also helps filter and remap interrupts
+from peripheral devices. Let's have a look at IOMMU advantages, disadvantages,
+and how it is implemented from the perspective of hardware and software.
+
+Advantages of IOMMU usage:
+
+  - One single contiguous virtual memory region can be mapped to multiple non-contiguous
+  physical memory regions. IOMMU can make a non-contiguous memory region appear
+  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
+  performance for the I/O device.
+  - Memory isolation and protection: device can only access memory regions that are
+  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
+  - Memory isolation allows safe device assignment to a virtual machine without
+  compromising host and other guest OSes.
+  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
+  - Support hardware interrupt remapping. It extends limited hardware interrupts
+  to software interrupts. Primary uses are the interrupt isolation and translation
+  between interrupt domains (E.g. ioapic vs x2apic on x86)
+
+Disadvantages of IOMMU usage:
+
+  - Latency in dynamic DMA mapping, translation overhead penalty.
+  - Host software has to maintain in-memory data structures for use by the IOMMU
+
+## In hardware
+
+IOMMU’s architecture is designed to accommodate a variety of system topologies.
+There can be multiple IOMMUs located at a variety of places in the system fabric.
+It can be placed at bridges inside one bus or at bridges between busses of the
+same or different types. In modern systems, IOMMU is commonly integrated with the
+PCIe root complex. It connects to the PCIe bus from downstream and to the system
+bus ( E.g. infinity fabric) from upstream. One example topology is shown below.
+
+![Example Platform Architecture](/img/iommu.png) 
+
+## In software
+
+The IOMMU is configured and controlled via two sets of registers. One in the PCI
+configuration space and another set mapped in system address space. Since the
+IOMMU appears to OS as a PCI function, it has a capability block in the PCI
+configuration space. Additionally, up to eight data structures are placed in the
+main system memory. These data structures contain information about the mapping
+of interrupts and memory for the usage of peripheral devices. Information about
+the device's memory access permissions is stored there too.
+
+## Enabling
+
+IOMMU is a generic name for, be Intel VT-d and AMD-Vi technologies. Make sure
+that your CPU supports one of these before you try to enable IOMMU.
```

Comment:
@Sakoram maybe it is worth to mention other implementations, especially IBM one since we entering the OpenPOWER ecosystem and ARM since we have a dedicated team working with that architecture.

Reply:
I added mentions about other implementation, but I couldn't find how to enable them in the kernel. I can't see a kernel option to enable it on PowerPC and ARM. [docs](https://www.kernel.org/doc/Documentation/admin-guide/kernel-parameters.txt ) say that iommu=on is meant only for x86.